velox/benchmarks/basic/LikeFunctionsBenchmark.cpp

```diff
@@ -0,0 +1,234 @@
+      case PatternKind::kExactlyN:
+        return std::string(inputString.size(), '_');
+      case PatternKind::kAtLeastN:
+        return generateRandomString(kWildcardCharacterSet);
+      case PatternKind::kPrefix: {
+        auto fixedPatternLength =
+            std::min(vector_size_t(inputString.size()), 10);
+        auto fixedPatternString = inputString.substr(0, fixedPatternLength);
+        return fixedPatternString + generateRandomString(kAnyWildcardCharacter);
+      }
+      case PatternKind::kSuffix: {
+        auto fixedPatternStartIdx =
+            std::max(vector_size_t(inputString.size() - 10), 0);
+        auto fixedPatternString = inputString.substr(fixedPatternStartIdx, 10);
+        return generateRandomString(kAnyWildcardCharacter) + fixedPatternString;
+      }
+      default:
+        return inputString;
+    }
+  }
+
+  const VectorPtr getTpchData(const TpchBenchmarkCase tpchCase) {
+    switch (tpchCase) {
+      case TpchBenchmarkCase::TpchQuery2:
+      case TpchBenchmarkCase::TpchQuery14:
+      case TpchBenchmarkCase::TpchQuery16Part: {
+        auto tpchPart = genTpchPart(FLAGS_num_rows);
+        return tpchPart->childAt(4);
+      }
+      case TpchBenchmarkCase::TpchQuery9:
+      case TpchBenchmarkCase::TpchQuery20: {
+        auto tpchPart = genTpchPart(FLAGS_num_rows);
+        return tpchPart->childAt(1);
+      }
+      case TpchBenchmarkCase::TpchQuery13: {
+        auto tpchOrders = genTpchOrders(FLAGS_num_rows);
+        return tpchOrders->childAt(8);
+      }
+      case TpchBenchmarkCase::TpchQuery16Supplier: {
+        auto tpchSupplier = genTpchSupplier(FLAGS_num_rows);
+        return tpchSupplier->childAt(6);
+      }
+      default:
+        VELOX_FAIL(fmt::format(
+            "Tpch data generation for case {} is not supported", tpchCase));
+    }
+  }
+
+  size_t run(const TpchBenchmarkCase tpchCase, const StringView patternString) {
+    folly::BenchmarkSuspender kSuspender;
+    const auto input = getTpchData(tpchCase);
+    const auto data = makeRowVector({input});
+    kSuspender.dismiss();
+
+    size_t cnt = 0;
+    for (auto i = 0; i < FLAGS_num_runs; i++) {
+      auto result = FunctionBaseTest::evaluate<SimpleVector<bool>>(
+          fmt::format("like(c0, '{}')", patternString), data);
+      cnt += result->size();
+    }
+    folly::doNotOptimizeAway(cnt);
+
+    return cnt;
+  }
+
+  size_t run(PatternKind patternKind) {
+    folly::BenchmarkSuspender kSuspender;
+    const auto input = inputFuzzer_->values()->as<StringView>();
+    auto patternString = generatePattern(patternKind, input[0].str());
+    std::vector<std::string> patternVector(FLAGS_vector_size, patternString);
+    const auto data = makeRowVector({inputFuzzer_});
+    kSuspender.dismiss();
+
+    size_t cnt = 0;
+    for (auto i = 0; i < FLAGS_num_runs; i++) {
+      auto result = FunctionBaseTest::evaluate<SimpleVector<bool>>(
+          fmt::format("like(c0, '{}')", patternString), data);
```

Comment:
thanks for the changing but are still compiling the expressions for every iteration , the expression compilation (parsing the string fmt::format("like(c0, '{}')", patternString) and compiling it to executable expressions should be pulled out to the suspended block and be done once.  

check /data/users/lsakka/fbsource/fbcode/velox/benchmarks/basic/Preproc.cpp
the function compile()
let me know if you need more pointers

Reply:
Thanks for the feedback, I have made changes using helper functions from FunctionBenchmarkBase.h, moving the compilation inside the suspender. I could optimize it further as in PreProc.cpp, maybe as follows:
```
  size_t run(const TpchBenchmarkCase tpchCase, const StringView patternString) {
    folly::BenchmarkSuspender kSuspender;
    const auto input = getTpchData(tpchCase);
    const auto data = makeRowVector({input});
    auto likeExpression = fmt::format("like(c0, '{}')", patternString);
    auto rowType = std::dynamic_pointer_cast<const RowType>(data->type());
    auto typedExpr = makeTypedExpr(likeExpression, rowType);
    SelectivityVector rows(data->size());
    std::vector<VectorPtr> results(1);
    
    std::shared_ptr<core::QueryCtx> queryCtx{core::QueryCtx::createForTest()};
    core::ExecCtx execCtx{FunctionBaseTest::pool(), queryCtx.get()};
    exec::ExprSet exprSet({typedExpr}, &execCtx);
    exec::EvalCtx evalCtx(&execCtx, &exprSet, data.get());
    kSuspender.dismiss();

    size_t cnt = 0;
    for (auto i = 0; i < FLAGS_num_runs; i++) {
      exprSet.eval(rows, evalCtx, results);
      cnt += results.size();
    }
    folly::doNotOptimizeAway(cnt);

    return cnt;
  }
```
Please let me know if the current changes are good or whether I should change as above. Thanks!